File: communication/sendMail.jl
#!/bin/env julia
#!/snap/bin/julia

# ADD Header :
# Date: Sat, 1 Jan 2000 12:00:59 +0200

import SMTPClient, Random, DBInterface, OteraEngine, TOML
using Dates

EMAIL_BODY_1ST_COMM_TEMPLATE_FILE = "template1stCommunication.html"
EMAIL_BODY_TEMPLATE_FILE = "templateMail.html"



ENV="dev"
conffile = "../.env.production"
if isfile(conffile)
	println("Production environment")
	ENV="prod"
else
	println("Development environment")
end


CURDIR=pwd()
include_path = ".."
if ENV=="dev"
	include_path = "../../openproduct-web-svelte4"
end
include(include_path*"/scripts/connect.jl")
dbConnection = get_connection(include_path)

if ENV=="dev"
	conffile = include_path*"/.env.local"
else
	conffile = include_path*"/.env.production"
end
println("Use configuration file : ", conffile)
conf = TOML.parsefile(conffile)
SMTP_HOST = conf["SMTP_HOST"]
SMTP_PORT = conf["SMTP_PORT"]
SMTP_PASS = conf["SMTP_PASS"]
SMTP_USER = conf["SMTP_USER"]

# using OpenProduct

function sendMail(to, subject, message, token)::Bool
	from = SMTP_USER
	fromC = "OpenProduct <"*from*">"
	nowdt = Dates.now()
	date = Dates.format(nowdt, RFC1123Format)*" +0100"
	premsg = "Date: "*date*"\r\n"*
		"List-Unsubscribe: <https://www.openproduct.fr/new/unsubscribe?mail="*to*"&token="*token*">\r\n"
	mime_msg = premsg*SMTPClient.get_mime_msg(HTML(replace(message, "\n"=>"\r\n")))
	println("MimeMsg:\n\n", premsg);
	rcpt = [to]
	body = SMTPClient.get_body(rcpt, fromC, subject, mime_msg)
	# println("Body:",String(take!(body)))
	url = "smtps://"*SMTP_HOST*":"*string(SMTP_PORT)
	opt = SMTPClient.SendOptions(isSSL=true, username=SMTP_USER, passwd=SMTP_PASS)
	resp = SMTPClient.send(url, rcpt, from, body, opt)
	if resp.code==250
		true
	else
		println("ERROR : fail send mail to ",to," : ",resp)
		false
	end
end

function generateToken()::String
	Random.seed!(0)
	Random.randstring(64)
end

function sendMail2AllProducers(subject, templateFilepath4Body)
	sql2 = """UPDATE producers
		SET has_send_mail=true
		WHERE email=\$1
		"""
	sqlUpdateSendEmail = DBInterface.prepare(dbConnection, sql2)
	sqlUpdateToken = nothing

	templateMailBody = OteraEngine.Template(templateFilepath4Body)

	sql = """SELECT p.company_name, coalesce(lastname, firstname , '') lastname, email, token
		FROM producers p 
		WHERE email is not null and email!=''
			AND send_email='ok'
			AND status in('actif', 'hs', 'to-check', 'unknown')
			AND has_send_mail=false
		ORDER BY ID
		LIMIT 10"""
	println("SQL:", sql);
	producers = DBInterface.execute(dbConnection, sql)
	myProducers = []
	for producer in producers
		push!(myProducers, (producer[1], producer[2], producer[3], producer[4]))
	end
	for producer in myProducers
		name = producer[1]
		lastname = producer[2]
		email = producer[3]
		token = producer[4]
		if ismissing(token) || token==""
			println("Generate token for ",email)
			token = generateToken()
			if isnothing(sqlUpdateToken)
				sql2 = """UPDATE producers 
					SET token=\$1
					WHERE email=\$2
					"""
				sqlUpdateToken = DBInterface.prepare(dbConnection, sql2)
			end
			DBInterface.execute(sqlUpdateToken, [token, email])
		end
		dictionary=Dict(
			Symbol("email")=> email,
			Symbol("lastname")=> lastname,
			Symbol("name")=> name,
			Symbol("token")=> token
		)
		message = templateMailBody(init=dictionary)
		slp = rand(1:240)
		println("Sleep ",slp," s.")
		sleep(slp) # Sleep between up to 4 minutes.
		# println("Send mail to ",email,".")
		ok = sendMail(email, subject, message, token)
		if ok
			DBInterface.execute(sqlUpdateSendEmail, [email])
			println("Sent message to ",email," (",name,", ",lastname,")")
		else
			println("Fail sent email")
		end
	end
end

function sendMailToNewProducers()
	sql2 = "UPDATE producers SET send_email='Ok' WHERE email=\$1"
	sqlUpdateSendEmail = DBInterface.prepare(dbConnection, sql2)
	sqlUpdateToken = nothing

	template1stCommunication = OteraEngine.Template(EMAIL_BODY_1ST_COMM_TEMPLATE_FILE)
	subject = ""

	sql = """SELECT name, coalesce(lastname,firstname,"") lastname, email, token
		FROM producers
		WHERE email is not null and email!=''
			AND send_email='unknown'
		ORDER BY ID
		"""
	println("SQL:", sql)
	producers = DBInterface.execute(dbConnection, sql)
	for producer in producers
		name = producer[1]
		lastname = producer[2]
		email = producer[3]
		token = producer[4]
		if ismissing(token) || token==""
			println("Generate token for ",email)
			token = generateToken()
			if isnothing(sqlUpdateToken)
				sql2 = """UPDATE producers
					SET token=\$1
					WHERE email=\$2
					"""
				sqlUpdateToken = DBInterface.prepare(dbConnection, sql2)
			end
			DBInterface.execute(sqlUpdateToken, [token, email])
		end
		dictionary=Dict(
			Symbol("email")=> email,
			Symbol("lastname")=> lastname,
			Symbol("name")=> name,
			Symbol("token")=> token
		)
		message = template1stCommunication(init=dictionary)
		ok = sendMail(email, subject, message, token)
		if ok
			DBInterface.execute(sqlUpdateSendEmail, [email])
			println("sent message to ",email," (",name,", ",lastname,")")
		end
	end
end

# sendMailToNewProducers()
# sendMailForAG()
sendMail2AllProducers(
	"OpenProduct évolue. Gérez désormais votre profil en direct",
	EMAIL_BODY_TEMPLATE_FILE
)
# sendMail("alberic.delacrochais@protonmail.com", "Test", "Un petit message de test")

File: communication/sendMail.py
#!/bin/env python3

from email.mime.text import MIMEText
import random
import os.path
import tomli
import psycopg
import smtplib, ssl
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import time

EMAIL_BODY_1ST_COMM_TEMPLATE_FILE = "template1stCommunication.html"
EMAIL_BODY_TEMPLATE_FILE = "templateMail.html"
EMAIL_SUBJECT = "OpenProduct évolue. Gérez désormais votre profil en direct"



ENV="dev"
conffile = "../.env.production"
if os.path.isfile(conffile):
	print("Production environment")
	ENV="prod"
else:
	conffile = "../../openproduct-web-svelte4/.env.local"
	print("Development environment")

DB_CONNECTION = None
def get_connection(root_path=".."):
	global DB_CONNECTION
	if DB_CONNECTION is None:
		# println("get_connection()", pwd())
		conffile = root_path+"/.env.production"
		if ENV=="dev":
			conffile = "../../openproduct-web-svelte4/.env.local"
		print("Use configuration file : ", conffile)
		with open(conffile, "rb") as f:
			conf = tomli.load(f)
			DATABASE_NAME = conf["DATABASE_NAME"]
			DATABASE_USER = conf["DATABASE_USER"]
			DATABASE_PASSWORD = conf["DATABASE_PASSWORD"]
			connection = psycopg.connect(dbname=DATABASE_NAME,
				user=DATABASE_USER, password=DATABASE_PASSWORD,
				host="localhost", port=5432)
			DB_CONNECTION = connection.cursor()
			print("Connected")
	return DB_CONNECTION
db_cnx = get_connection()

SMTP_USER = None
SMTP_HOST = None
SMTP_PORT = None
SMTP_PASS = None
with open(conffile, "rb") as f:
	conf = tomli.load(f)
	SMTP_USER = conf["SMTP_USER"]
	SMTP_HOST = conf["SMTP_HOST"]
	SMTP_PORT = conf["SMTP_PORT"]
	SMTP_PASS = conf["SMTP_PASS"]
	
def sendMail(to, subject, body, token):
	global SMTP_USER
	fromC = "OpenProduct <"+SMTP_USER+">"
	premsg = "List-Unsubscribe: <https://www.openproduct.fr/new/unsubscribe?mail="+to+"&token="+token+">\r\n"
	context = ssl.create_default_context()
	message = MIMEMultipart("alternative")
	message["Subject"] = subject
	message["From"] = fromC
	message["To"] = to
	message.attach(MIMEText(body, "html"))
	message.add_header('List-Unsubscribe', '<https://www.openproduct.fr/new/unsubscribe?mail='+to+'token='+token+'>')
	context = ssl.create_default_context()
	with smtplib.SMTP_SSL(SMTP_HOST, SMTP_PORT, context=context) as server:
		server.login(SMTP_USER, SMTP_PASS)
		ok = server.sendmail(SMTP_USER, to, message.as_string())
		return True
	return False

# Query for all email
db_cnx.execute(
	"""SELECT p.company_name, coalesce(lastname, firstname , '') lastname, email, token
		FROM producers p 
		WHERE email is not null and email!=''
			AND send_email='ok'
			AND status in('actif', 'to-check', 'unknown')
			AND has_send_mail=false
		ORDER BY ID
		LIMIT 10
	"""
)
producers = db_cnx.fetchall()
print(producers)

# sendMail("alberic.delacrochais@protonmail.com", EMAIL_SUBJECT, "Un test", "ugiofnkjulblfu")
# producers = [('OpenHomeSystem', 'Albéric', 'alberic.delacrochais@protonmail.com', '5b76db3e3f1d426ede4fce655844366b')]

from jinja2 import Template
with open(EMAIL_BODY_TEMPLATE_FILE, 'r') as f:
	template = Template(f.read())

	for producer in producers:
		email = producer[2]
		token = producer[3]
		context = {
			'email': email,
			'lastname': producer[1],
			'name': producer[0],
			'token': token
		}
		body = template.render(context)
		slp =  random.randrange(1, 240)
		print("Sleep ",slp," s.")
		time.sleep(slp)
		if sendMail(email, EMAIL_SUBJECT, body, token):
			sql = "UPDATE producers SET has_send_mail=True WHERE email='"+email+"'"
			db_cnx.execute(sql)
			print("Sent message to ",email," successfully.")
		else:
			print("Fail send mail to ",email)

